eax/src/online.rs

```diff
@@ -0,0 +1,432 @@
+/// [`Eax`]: ../struct.Eax.html
+/// [`Decrypt`]: struct.Decrypt.html
+/// [`finish`]: #method.finish
+pub struct Eax<Cipher, Op>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+    Op: CipherOp,
+{
+    imp: EaxImpl<Cipher>,
+    /// Denotes whether this stream is used for encryption or decryption.
+    marker: PhantomData<Op>,
+}
+
+impl<Cipher, Op> Eax<Cipher, Op>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+    Op: CipherOp,
+{
+    /// Creates a stateful EAX instance that is capable of processing both
+    /// the associated data and the plaintext in an "on-line" fashion.
+    pub fn with_key_and_nonce(key: &Key<Cipher>, nonce: &Nonce<Cipher::BlockSize>) -> Self {
+        let imp = EaxImpl::<Cipher>::with_key_and_nonce(key, nonce);
+
+        Self {
+            imp,
+            marker: PhantomData,
+        }
+    }
+
+    /// Process the associated data (AD).
+    #[inline]
+    pub fn update_assoc(&mut self, aad: &[u8]) {
+        self.imp.update_assoc(aad);
+    }
+
+    /// Derives the tag from the encrypted/decrypted message so far.
+    ///
+    /// If the encryption/decryption operation is finished, [`finish`] method
+    /// *must* be called instead.
+    ///
+    ///[`finish`]: #method.finish
+    #[inline]
+    pub fn tag_clone(&self) -> Tag {
+        self.imp.tag_clone()
+    }
+}
+
+impl<Cipher> Eax<Cipher, Encrypt>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+{
+    /// Applies encryption to the plaintext.
+    #[inline]
+    pub fn encrypt(&mut self, msg: &mut [u8]) {
+        self.imp.encrypt(msg)
+    }
+
+    /// Finishes the encryption stream, returning the derived tag.
+    ///
+    /// This *must* be called after the stream encryption is finished.
+    #[must_use = "tag must be saved to later verify decrypted data"]
+    #[inline]
+    pub fn finish(self) -> Tag {
+        self.imp.tag()
+    }
+}
+
+impl<Cipher> Eax<Cipher, Decrypt>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+{
+    /// Applies decryption to the ciphertext **without** verifying the
+    /// authenticity of decrypted message.
+    ///
+    /// To correctly verify the authenticity, use the [`finish`] associated
+    /// function.
+    ///
+    /// [`finish`]: #method.finish
+    #[inline]
+    pub fn decrypt_unauthenticated(&mut self, msg: &mut [u8]) {
+        self.imp.decrypt(msg)
+    }
+
+    /// Finishes the decryption stream, verifying whether the associated and
+    /// decrypted data stream has not been tampered with.
+    ///
+    /// This *must* be called after the stream decryption is finished.
+    #[must_use = "decrypted data stream must be verified for authenticity"]
```

Comment:
`Result` is already `#[must_use]` so I don't think this accomplishes anything?

I think it'd also be good to talk about the breakages that failure to call `finish` might cause, but those should also be mentioned in `decrypt_unauthenticated`.

Reply:
That's to add the specific note. Do you want me to remove it?

Reply:
If that actually works, we can keep it. Does it?

Reply:
Yep, simply calling
```rust
cipher.finish(&tag);
```
yields the following warning:
```
warning: unused return value of `eax::online::Eax::<Cipher, eax::online::Decrypt>::finish` that must be used
  --> eax/tests/lol.rs:30:5
   |
30 |     cipher.finish(&tag);
   |     ^^^^^^^^^^^^^^^^^^^^
   |
   = note: decrypted data stream must be verified for authenticity
```

Do you want me to keep it or reword it?

Reply:
That's fine